### ISSUE TITLE
Allow to specify the value of the `trainer/global_step` metric on W&B during validation

### ISSUE THREAD

**Issue by andreapesare** [Lightning-AI/lightning#14892]
## 🚀 Feature

Allow to specify the value of the `trainer/global_step` metric on W&B even during validation, as it is possible during training.

### Motivation

The default value for the `trainer/global_step` metric during validation is defined as the "total validation step":

https://github.com/Lightning-AI/lightning/blob/e7ad151c311890c931aa5ff3df3672fd21fd351c/src/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py#L152

whereas during training the default value is `trainer.fit_loop.epoch_loop._batches_that_stepped`, even if there are some inconsistencies (see #13752).

The fact that the behaviour of `trainer/global_step` is different when switching from training to validation, generates graphs like this on W&B:

<img width="1342" alt="image" src="https://user-images.githubusercontent.com/93650955/192155587-625e8bfc-2885-4af9-afd2-67b4f74f69eb.png">

as well as the fact that W&B complains when trying to use `trainer/global_step` as `x` axis because it's not monotonically increasing:

<img width="518" alt="image" src="https://user-images.githubusercontent.com/93650955/192155737-b03801e3-4a1c-4f74-b88e-e0005c61939e.png">

For training it's possible to set a custom value for `training/global_step` by using `self.log("step", ...)` during the training step. For validation, there's no such possibility.

I understand that if one logs using `on_step=True` during validation, it's important to keep track of the different validation steps. But if one logs using `on_step=False` and `on_epoch=True`, it would be useful to link the val metric to the current training step instead of the total validation step.

### Pitch

One can set a custom value for `training/global_step` even during validation.

### Alternatives

Another possibility could be to log a different metric for validation, e.g. `trainer/validation_global_step` and to use that as default x axis as it is done with `trainer/global_step` here:

https://github.com/Lightning-AI/lightning/blob/6df6deac8fdf0c44b509d531441b4f7e446de6b4/src/pytorch_lightning/loggers/wandb.py#L396-L397

**Comment by andreapesare** [comment]
My bad, the issue persists only when logging with `on_step=True` and thus is a minor problem.

When logging with `on_epoch=True` (which is the default during validation), the behaviour is the same as during training, i.e. it uses the current training step:

https://github.com/Lightning-AI/lightning/blob/759e89df21684277da996dd25781dee7f21e1b1c/src/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L136-L142

However, I believe that the alternative solution outlined above could be valid to avoid some confusion.

**Comment by manangoel99** [comment]
Hi @andreapesare ! Thanks  for pointing this out. Would it be possible to get some code on how to reproduce the global_step plot that you showed above?

**Comment by andreapesare** [comment]
Hi @manangoel99 ! This code should reproduce a similar plot, when using `step` as x axis:

```python
import os
from typing import Optional

import pytorch_lightning as pl
import torch
import torch.nn.functional as F
from pytorch_lightning.loggers.wandb import WandbLogger
from pytorch_lightning.utilities.types import STEP_OUTPUT
from torch import nn
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST


class LitModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.log("train/loss", loss)
        return loss

    def validation_step(self, batch, batch_idx) -> Optional[STEP_OUTPUT]:
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.log("val/loss", loss, on_step=True, on_epoch=False)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)


train_loader = DataLoader(
    MNIST(os.getcwd(), download=True, train=True, transform=transforms.ToTensor())
)
val_loader = DataLoader(
    MNIST(os.getcwd(), download=True, train=False, transform=transforms.ToTensor())
)
logger = WandbLogger(project="mnist-example")
trainer = pl.Trainer(
    max_epochs=10,
    limit_train_batches=10,
    limit_val_batches=10,
    log_every_n_steps=1,
    logger=logger,
)
model = LitModel()

trainer.fit(model, train_dataloaders=train_loader, val_dataloaders=val_loader)
```

Here the plot that I get running the code above:
![image](https://user-images.githubusercontent.com/93650955/192488006-866eaed2-5405-4d4a-8c39-36b668901346.png)

**Comment by manangoel99** [comment]
Thanks @andreapesare ! Looking into it
